### PR TITLE
Parser and logger

### PR DESCRIPTION
**Pull request by cogitantium**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+        edg::distributed_certain_zero(EmptyGraph {}, 0, num_cpus::get() as u64);
+    }
+}
+
+/// Define and parse command line arguments
+fn parse() -> ArgMatches<'static> {
+    App::new("OnTheFlyATL")
+        .version("0.1.0")
+        .author("d702e20 <d702e20@cs.aau.dk>")
+        .arg(Arg::with_name("proposition")
```
Comment (lars4509) [inline src/main.rs:59]: I think you mean 'formula'?

File: src/main.rs [hunk 2]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+            .short("p")
+            .long("prop")
+            .env("PROPOSITION")
+            .help("The proposition to check for"))
```
Comment (NicEastvillage) [inline src/main.rs:63]: ATL formula, not a proposition

**Review by lars4509 — CHANGES_REQUESTED** [review]
Overall looks fine, but i think you mean ATL formula, instead of proposition

**Review by NicEastvillage — APPROVED** [review]

File: src/main.rs [hunk 3]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+            .short("o")
+            .long("log-level")
+            .env("LOG_LEVEL")
+            .default_value("info")
+            .help("{error, warn, info, debug, trace, off}"))
+        .arg(Arg::with_name("log-path")
+            .short("g")
+            .long("log-path")
+            .env("LOG_PATH")
+            .default_value("model-checker.log")
```
Comment (jenrik) [inline src/main.rs:89]: how do we disable log files? Can we make it default to not writing any log files?

File: src/main.rs [hunk 4]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+        .arg(Arg::with_name("json_formula")
+            .short("r")
+            .long("jsonformula")
+            .env("JSON_FORMULA")
+            .help("The json to generate formula from"))
+        .arg(Arg::with_name("log-level")
+            .short("o")
+            .long("log-level")
+            .env("LOG_LEVEL")
+            .default_value("info")
```
Comment (jenrik) [inline src/main.rs:83]: the UNIX philosophy is to not print anything unless there is an error or the user requested it

```suggestion
            .default_value("warn")
```

File: src/main.rs [hunk 5]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+        .arg(Arg::with_name("json_model")
+            .short("j")
+            .long("json")
+            .env("INPUT_JSON")
+            .help("The json to generate model from"))
+        .arg(Arg::with_name("json_formula")
+            .short("r")
+            .long("jsonformula")
+            .env("JSON_FORMULA")
+            .help("The json to generate formula from"))
```
Comment (jenrik) [inline src/main.rs:78]:
```suggestion
            .help("Path to a JSON formatted formula that will be checked against the model"))
```

File: src/main.rs [hunk 6]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+        .arg(Arg::with_name("input_file")
+            .short("i")
+            .long("input")
+            .env("INPUT_FILE")
+            .help("The input file to generate model from"))
+        .arg(Arg::with_name("json_model")
+            .short("j")
+            .long("json")
+            .env("INPUT_JSON")
+            .help("The json to generate model from"))
```
Comment (jenrik) [inline src/main.rs:73]:
```suggestion
            .help("Path to the model in JSON format"))
```

File: src/main.rs (hunk 2 above)
Comment (jenrik) [inline src/main.rs]:
```suggestion
```

File: src/main.rs (hunk 1 above)
Comment (cogitantium) [reply src/main.rs]: Yes, of course - fixed in ab78e01

File: src/main.rs (hunk 2 above)
Comment (cogitantium) [reply src/main.rs]: Yes, of course - fixed in ab78e01

**Review by lars4509 — APPROVED** [review]

File: src/main.rs [hunk 7]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+    if let Ok(_handle) = log4rs::init_config(get_log4rs_config(
+        args.value_of("log-path").unwrap(),
+        match args.value_of("log-level").unwrap().to_lowercase().as_str() {
+            "error" => LevelFilter::Error,
+            "warn" => LevelFilter::Warn,
+            "info" => LevelFilter::Info,
+            "debug" => LevelFilter::Debug,
+            "trace" => LevelFilter::Trace,
+            "off" => LevelFilter::Off,
+            _ => LevelFilter::Off,
```
Comment (jenrik) [inline src/main.rs:45]: We should probably fail with an informative error message instead of silently accepting wrong input

**Review by jenrik — CHANGES_REQUESTED** [review]
The language of the help messages is inconsistent, and they should be reworded in my opinion.

The Tracing framework might be a better alternative for logging and getting diagnostics, see https://tracing.rs/tracing/

File: src/main.rs (hunk 4 above)
Comment (NicEastvillage) [reply src/main.rs:83]: updut

File: src/main.rs [hunk 8]
```diff
@@ -20,5 +30,90 @@ impl edg::ExtendedDependencyGraph<i32> for EmptyGraph {
...
+
+        edg::distributed_certain_zero(EmptyGraph {}, 0, num_cpus::get() as u64);
+    }
+}
+
+/// Define and parse command line arguments
+fn parse() -> ArgMatches<'static> {
+    App::new("OnTheFlyATL")
+        .version("0.1.0")
+        .author("d702e20 <d702e20@cs.aau.dk>")
```
[the remark this answers is not in the document]
Comment (cogitantium) [reply src/main.rs:58]: Fixed in 34082d9

File: src/main.rs (hunk 3 above)
Comment (cogitantium) [reply src/main.rs:89]: Added argument to disable and made it default in d69b7c2

File: src/main.rs (hunk 4 above)
Comment (cogitantium) [reply src/main.rs:83]: Agreed, this was default in d502e19 project. Fixed in d69b7c2

File: src/main.rs (hunk 5 above)
Comment (cogitantium) [reply src/main.rs:78]: Applied in d69b7c2

File: src/main.rs (hunk 6 above)
Comment (cogitantium) [reply src/main.rs:73]: Applied in d69b7c2

File: src/main.rs (hunk 7 above)
Comment (cogitantium) [reply src/main.rs:45]: Good idea, applied in d69b7c2

**Pull request merged**